CI: group tests in stages

psychocoderHPC opened the pull request:
- Group tests in stages to reduce the CI pressure. In cases where a stage is failing it makes no sense to run all others.
- Additionally, the second commit set the environment variable `ALPAKA_CI` to avoid running the `headerCheck` if not explicitly activated.
- Disable `randomCells2D` for the OpenACC backend because of invalid memory access. The reason for the issue should be investigated independently to avoid blocking other PRs.

Per YAML file I incremented the stage id, I have not balanced the stages by hand because this would make it hard to keep an overview.

Writing everything by hand should be only a workaround we need asap a generator https://github.com/alpaka-group/alpaka/issues/1517#issuecomment-988044352

psychocoderHPC committed:
CI: group tests in stages

Group tests in stages to reduce the CI pressure. In cases where a istage is
failing it makes no sense to run all others.

psychocoderHPC committed:
CI: define `ALPAKA_CI` for gitlab

Define `ALPAKA_CI` for the HZDR gitlab CI. This will deactivate some
tests, e.g. `headerCheck`

bernhardmgruber commented:
I see `nvhpc` in the last 8th stage of the pipeline and wonder whether we should have it earlier to avoid running 7 stages to then fail in the 8th because of a very unique compiler.

psychocoderHPC commented:
> I see `nvhpc` in the last 8th stage of the pipeline and wonder whether we should have it earlier to avoid running 7 stages to then fail in the 8th because of a very unique compiler.

I also thought about it, if so we should reorder the yaml file stay with simple indexing within the file. We could do it in a separate PR but should finish first our open PRs.

psychocoderHPC committed:
CI: OpenACC disable random2D example

The example random2D is producing invalid memory accesses when running
in the CI. UNtil we found the reason this example will be disabled.

bernhardmgruber commented on example/CMakeLists.txt:
```diff
@@ -36,6 +36,9 @@ add_subdirectory("kernelSpecialization/")
 add_subdirectory("openMPSchedule/")
 add_subdirectory("randomStrategies/")
-add_subdirectory("randomCells2D/")
+# This example is producing invalid memory access during CI tests with the OpenACC backend.
+if(NOT ALPAKA_CI OR NOT ALPAKA_ACC_ANY_BT_OACC_ENABLE)
+    add_subdirectory("randomCells2D/")
+endif()
```
This needs to be fixed at some point (discussed offline): #1540.

psychocoderHPC replied:
Thanks for opening the issue. It was to late yesterday night to do it.

bernhardmgruber commented on .gitlab-ci.yml:
```diff
@@ -1,3 +1,14 @@
+stages:
+  - stage0
+  - stage1
+  - stage2
+  - stage3
+  - stage4
+  - stage5
+  - stage6
+  - stage7
+  - stage8
```
I am worried that the pipeline is too deep and CI builds take too long. But we will see.
